py/desitarget/brightstar.py

```diff
@@ -453,22 +660,208 @@ def is_in_bright_star(targs,starmask):
+    #ADM determine the indexes of the targets that have a TARGETID in matches
+    w_mask = [ index for index, item in enumerate(targetid) if item in matches ]
+
+    #ADM w_mask now contains the target indices that match to a bright star mask on TARGETID
+    is_mask[w_mask] = 'True'
+
+    return is_mask
+
+def generate_safe_locations(starmask,Npersqdeg):
+    """Given a bright star mask, generate SAFE (BADSKY) locations at its periphery
+
+    Parameters
+    ----------
+    starmask : :class:`recarray`
+        A recarray containing a bright star mask as made by :mod:`desitarget.brightstar.make_bright_star_mask`
+    npersqdeg : :class:`int`
+        The number of safe locations to generate per square degree of each mask
+
+    Returns
+    -------
+    ra : array_like. 
+        The Right Ascensions of the SAFE (BADSKY) locations
+    dec : array_like. 
+        The Declinations of the SAFE (BADSKY) locations
+
+    Notes
+    -----
+        - See the Tech Note at https://desi.lbl.gov/DocDB/cgi-bin/private/ShowDocument?docid=2346 for more details
+    """
+    
+    #ADM the radius of each mask in degrees with a 0.1% kick to get things beyond the mask edges
+    radius = 1.001*starmask["IN_RADIUS"]/60.
+
+    #ADM determine the area of each mask
+    area = cap_area(radius)
+
+    #ADM determine the number of SAFE locations to assign to each
+    #ADM mask given the passed number of locations per sq. deg.
+    Nsafe = np.ceil(area*Npersqdeg).astype('i')
+
+    #ADM determine Nsafe Dec offsets equally spaced around the perimeter for each mask
+    offdec = [ rad*np.sin(np.arange(ns)*2*np.pi/ns) for ns, rad in zip(Nsafe,radius) ]
+
+    #ADM use offsets to determine DEC positions
+    dec = starmask["DEC"] + offdec
+
+    #ADM determine the offsets in RA at these Decs given the mask center Dec
+    offrapos =  [ sphere_circle_ra_off(th,cen,declocs) for th,cen,declocs in zip(radius,starmask["DEC"],dec) ]
+
+    #ADM determine which of the RA offsets are in the positive direction
+    sign = [ np.sign(np.cos(np.arange(ns)*2*np.pi/ns)) for ns in Nsafe ]
+
+    #ADM determine the RA offsets with the appropriate sign and add them to the RA of each mask
+    offra = [ o*s for o,s in zip(offrapos,sign) ]
+    ra = starmask["RA"] + offra
+
+    #ADM have to turn the generated locations into 1-D arrays before returning them
+    return np.hstack(ra), np.hstack(dec)
+
+
+def append_safe_targets(targs,starmask,drstring=None,drbricks=None):
+    """Append targets at SAFE (BADSKY) locations to target list, set bits in TARGETID and DESI_TARGET
+
+    Parameters
+    ----------
+    targs : :class:`recarray`
+        A recarray of targets as made by desitarget.cuts.select_targets
+    starmask : :class:`recarray`
+        A recarray containing a bright star mask as made by desitarget.brightstar.make_bright_star_mask
+    drstring : :class:`str`, optional
+        The imaging data release for the targets as a 9-string (i.e. 'dr3      ') to update TARGETID
```

Comment:
This is purely because we specified this as a 9-string here:

https://desi.lbl.gov/DocDB/cgi-bin/private/RetrieveFile?docid=2348;filename=targetID.pdf;version=1

(see the final section). We could pass this as an integer here, if you'd prefer, and convert it to a string/pad it for other purposes? We'll need both, though (one for the `TARGETCAT` hash and one for bit-encoding.